### PR TITLE
make modifications for transferAllocation function

### PR DESCRIPTION
**Pull request by divine-comedian**
change logic of cancelAllocation to add allocation balance instead of overwriting it.

### CODE REVIEW

File: contracts/ModifiedTokenDistro.sol
Comment (aminlatifi) [inline contracts/ModifiedTokenDistro.sol]: SafeMath is not needed, since solidity v0.8 overflow and underflow checks are implemented on language level

File: contracts/ModifiedTokenDistro.sol
Comment (aminlatifi) [inline contracts/ModifiedTokenDistro.sol]: Likewise

File: contracts/ModifiedTokenDistro.sol
Comment (aminlatifi) [inline contracts/ModifiedTokenDistro.sol]: I noticed you renamed the function along with removing this check
```solidity
        require(cancelable, 'TokenDistro::cancelAllocation: NOT_CANCELABLE');
```
This change will make the `cancelable` property unused, and we cannot remove it unless we fill it with a dummy variable for memory safety reasons.
Do you still think it's better to rename it and remove the check?

File: contracts/ModifiedTokenDistro.sol
Comment (aminlatifi) [inline contracts/ModifiedTokenDistro.sol]: The arithmetic can be done without use of SafeMath, like what has happened in `changeAddress` function. Plus, I think since these two methods doing similar logic, let's define an internal method which transfers balance of an account to another and use that inside these two.

**Review by aminlatifi — CHANGES_REQUESTED** [review]

File: contracts/ModifiedTokenDistro.sol
Comment (divine-comedian) [reply contracts/ModifiedTokenDistro.sol]: I added back the check - Do you think there will be consequences for renaming the function for memory safety?

**Comment by divine-comedian** [comment]
@jainkrati  - this one needs a review to unblock me

File: contracts/ModifiedTokenDistro.sol
Comment (jainkrati) [inline contracts/ModifiedTokenDistro.sol]: before this we should ensure
balances[address(this)].allocatedTokens > amount

File: contracts/ModifiedTokenDistro.sol
Comment (jainkrati) [inline contracts/ModifiedTokenDistro.sol]: before this lets check

balances[msg.sender].allocatedTokens > amount

File: contracts/ModifiedTokenDistro.sol
Comment (jainkrati) [inline contracts/ModifiedTokenDistro.sol]: lets cache array length before for loop and then use it. will save gas.


uint256 length = recipients.length;
for (uint256 i = 0; i < length; i++) {

File: contracts/ModifiedTokenDistro.sol
Comment (jainkrati) [inline contracts/ModifiedTokenDistro.sol]: just to be on safe side, pls check that duration > 0

File: contracts/ModifiedTokenDistro.sol
Comment (jainkrati) [inline contracts/ModifiedTokenDistro.sol]: totalTokens > 0 check before this line pls

File: contracts/ModifiedTokenDistro.sol
Comment (jainkrati) [inline contracts/ModifiedTokenDistro.sol]: before this pls check 
unlockedAmount > balances[recipient].claimed

File: contracts/ModifiedTokenDistro.sol
Comment (jainkrati) [inline contracts/ModifiedTokenDistro.sol]: is there a possibility that balances[prevRecipient].allocatedTokens == balances[prevRecipient].claimed ? If yes lets check it before.

File: test/ModifyTokenDistro.t.sol
Comment (jainkrati) [inline test/ModifyTokenDistro.t.sol]: as a good practice lets remove commented code. we can always access it from github history when needed.

**Review by jainkrati — CHANGES_REQUESTED** [review]
Pls dm me once comments are resolved

**Review by jainkrati — APPROVED** [review]

File: contracts/ModifiedTokenDistro.sol
Comment (jainkrati) [reply contracts/ModifiedTokenDistro.sol]: Renaming might result in fallback call when function is invoked. @sembrestels would appreciate your comment on this.

Comment (jainkrati) [reply contracts/ModifiedTokenDistro.sol]: Please add this comment above the function definition

// this function should only be called manually

**Review by jainkrati — APPROVED** [review]
Looks good. just 1 comment

**Comment by jainkrati** [comment]
Also can we make the Slither Analysis pass before we merge?

File: contracts/ModifiedTokenDistro.sol
Comment (aminlatifi) [inline contracts/ModifiedTokenDistro.sol]: It's not needed. In case that condition is met, the subtraction below will underflow and tx will revert
```solidity
        balances[address(this)].allocatedTokens = balances[address(this)].allocatedTokens - amount;
```

File: contracts/ModifiedTokenDistro.sol
Comment (aminlatifi) [inline contracts/ModifiedTokenDistro.sol]: You don't want to follow the same convention and define this method in `IDistro` as well?

File: contracts/ModifiedTokenDistro.sol
Comment (aminlatifi) [inline contracts/ModifiedTokenDistro.sol]: We don't need this check
`balances[prevRecipient].allocatedTokens != balances[prevRecipient].claimed`

File: contracts/ModifiedTokenDistro.sol
Comment (aminlatifi) [inline contracts/ModifiedTokenDistro.sol]: Likewise

File: contracts/ModifiedTokenDistro.sol
Comment (aminlatifi) [inline contracts/ModifiedTokenDistro.sol]: Have you tested extracting this constant variable to save gas for us? If yes it's okay. Otherwise, I won't change it even if it's trivial. 
It will be so bad for us if this change doesn't make any improvement and even increase the gas cost.

File: contracts/ModifiedTokenDistro.sol
Comment (aminlatifi) [inline contracts/ModifiedTokenDistro.sol]: No need to do this. In case the duration be zero, this division will cause tx revert
```
return initialAmount + (deltaTime * lockedAmount) / duration;
```

File: contracts/ModifiedTokenDistro.sol
Comment (aminlatifi) [inline contracts/ModifiedTokenDistro.sol]: Likewise, no need for this.

**Review by aminlatifi — CHANGES_REQUESTED** [review]

File: contracts/ModifiedTokenDistro.sol
Comment (jainkrati) [reply contracts/ModifiedTokenDistro.sol]: gas fee is same for both approaches. With require() you get a better error description. Tx reverst in both. your call guys.

File: contracts/ModifiedTokenDistro.sol
Comment (divine-comedian) [reply contracts/ModifiedTokenDistro.sol]: Done!

File: contracts/ModifiedTokenDistro.sol
Comment (jainkrati) [reply contracts/ModifiedTokenDistro.sol]: It is a proven strategy for saving gas. request you to take a look at point 8 in https://coinsbench.com/comprehensive-guide-tips-and-tricks-for-gas-optimization-in-solidity-5380db734404 and https://medium.com/@vgabrielmarian21/solidity-caching-length-in-for-loops-d02402628741

Comment (aminlatifi) [reply contracts/ModifiedTokenDistro.sol]: Good point, thanks @jainkrati

File: contracts/ModifiedTokenDistro.sol
Comment (aminlatifi) [inline contracts/ModifiedTokenDistro.sol]:
```suggestion
    function sendPraiseRewards(address[] memory recipients, uint256[] memory amounts) external override {
```

**Review by aminlatifi — CHANGES_REQUESTED** [review]
Thanks @divine-comedian 
Looks good, just a simple small change suggested

**Comment by aminlatifi** [comment]
@divine-comedian @jainkrati I did move the changes under the same contract name TokenDistro, since it's not trivial to change the contract name in the upgrade process.
Also, made #43 to push latest version of production to `develop` branch, I think it's better to merge it first to the develop and then do the final comparison.
BTW, I think the code is in good state now, just deleted an extra empty line.

**Comment by aminlatifi** [comment]
Thanks @divine-comedian for fixing the slither action, I just did a little change to make it work.

**Comment by divine-comedian** [comment]
@aminlatifi thanks! I reviewed this one, can you approve it then?

**Review by aminlatifi — APPROVED** [review]

**Comment by aminlatifi** [comment]
> @aminlatifi thanks! I reviewed this one, can you approve it then?

Approved

**Pull request merged**
